scalanet/src/io/iohk/scalanet/peergroup/dynamictls/DynamicTLSPeerGroup.scala

```diff
@@ -0,0 +1,395 @@
+  * that are not instances of TLSPeerGroup.
+  *
+  * @param config bind address etc. See the companion object.
+  * @param codec  a decco codec for reading writing messages to NIO ByteBuffer.
+  * @tparam M the message type.
+  */
+class DynamicTLSPeerGroup[M](val config: Config)(
+    implicit codec: StreamCodec[M],
+    scheduler: Scheduler
+) extends TerminalPeerGroup[PeerInfo, M]() {
+
+  private val log = LoggerFactory.getLogger(getClass)
+
+  private val sslServerCtx: SslContext = DynamicTLSPeerGroupUtils.buildCustomSSlContext(SSLContextForSever, config)
+
+  private val serverSubject = ConnectableSubject[ServerEvent[PeerInfo, M]]()
+
+  private val workerGroup = new NioEventLoopGroup()
+
+  private val clientBootstrap = new Bootstrap()
+    .group(workerGroup)
+    .channel(classOf[NioSocketChannel])
+    .option[java.lang.Boolean](ChannelOption.SO_KEEPALIVE, true)
+    .option[RecvByteBufAllocator](ChannelOption.RCVBUF_ALLOCATOR, new DefaultMaxBytesRecvByteBufAllocator)
+    .option[Integer](ChannelOption.CONNECT_TIMEOUT_MILLIS, 10000)
+
+  private val serverBootstrap = new ServerBootstrap()
+    .group(workerGroup)
+    .handler(new LoggingHandler(LogLevel.DEBUG))
+    .channel(classOf[NioServerSocketChannel])
+    .childHandler(new ChannelInitializer[SocketChannel]() {
+      override def initChannel(ch: SocketChannel): Unit = {
+        new ServerChannelBuilder[M](serverSubject, ch, sslServerCtx, codec)
+        log.info(s"$processAddress received inbound from ${ch.remoteAddress()}.")
+      }
+    })
+    .option[Integer](ChannelOption.SO_BACKLOG, 128)
+    .option[RecvByteBufAllocator](ChannelOption.RCVBUF_ALLOCATOR, new DefaultMaxBytesRecvByteBufAllocator)
+    .childOption[java.lang.Boolean](ChannelOption.SO_KEEPALIVE, true)
+
+  private lazy val serverBind: ChannelFuture = serverBootstrap.bind(config.bindAddress)
+
+  override def initialize(): Task[Unit] =
+    toTask(serverBind).map(_ => log.info(s"Server bound to address ${config.bindAddress}")).onErrorRecoverWith {
+      case NonFatal(e) => Task.raiseError(InitializationError(e.getMessage, e.getCause))
+    }
+
+  override def processAddress: PeerInfo = config.peerInfo
+
+  override def client(to: PeerInfo): Task[Channel[PeerInfo, M]] = {
+    // Creating new ssl context for each client is necessary, as this is only reliable way to pass peerInfo to TrustManager
+    // which takes care of validating certificates and server node id.
+    // Using Netty SSLEngine.getSession.putValue does not work as expected as until successfulhandshake there is no separate
+    // session for each connection.
+    new ClientChannelImpl[M](
+      to,
+      clientBootstrap,
+      DynamicTLSPeerGroupUtils.buildCustomSSlContext(SSLContextForClient(to), config),
+      codec
+    ).initialize
+  }
+
+  override def server(): ConnectableObservable[ServerEvent[PeerInfo, M]] = serverSubject
+
+  override def shutdown(): Task[Unit] = {
+    serverSubject.onComplete()
```

Comment:
I don't know the API but `serverSubject.onComplete()` looks strange

Reply:
it is monix way to tell all subscribers that the stream have ended.  It means that anyone listenning on server events on `serverSubject` will be informed that there will be no more elements.